src/wp-includes/media.php

```diff
@@ -5651,104 +5624,154 @@ function wp_get_loading_optimization_attributes( $tag_name, $attr, $context ) {
+					break;
+				}
+			// phpcs:ignore Generic.WhiteSpace.ScopeIndent.Incorrect
+			// Consider elements before the loop as being in viewport.
+			case 'wp_get_attachment_image':
+			case 'widget_media_image':
+				if (
+					// Only apply for main query but before the loop.
+					$wp_query->before_loop && $wp_query->is_main_query()
+					/*
+					 * Any image before the loop, but after the header has started should not be lazy-loaded,
+					 * except when the footer has already started which can happen when the current template
+					 * does not include any loop.
+					 */
+					&& did_action( 'get_header' ) && ! did_action( 'get_footer' )
+				) {
+					$in_viewport          = true;
+					$maybe_increase_count = true;
+				}
+				break;
 		}
 	}
 
 	/*
-	 * The first elements in 'the_content' or 'the_post_thumbnail' should not be lazy-loaded,
-	 * as they are likely above the fold. Shortcodes are processed after content images, so if
-	 * thresholds haven't already been met, apply the same logic to those as well.
+	 * If the element is in the viewport, potentially add `fetchpriority` with
+	 * a value of "high". Otherwise (element not in viewport or viewport is
+	 * unknown), add `loading` with a value of "lazy".
 	 */
-	if ( 'the_content' === $context || 'the_post_thumbnail' === $context || 'do_shortcode' === $context ) {
-		// Only elements within the main query loop have special handling.
-		if ( is_admin() || ! in_the_loop() || ! is_main_query() ) {
-			$loading_attrs['loading'] = 'lazy';
-			return $postprocess( $loading_attrs, false );
-		}
+	if ( $in_viewport ) {
+		$loading_attrs = wp_maybe_add_fetchpriority_high_attr( $loading_attrs, $tag_name, $attr );
+	} else { // `$in_viewport` is either `false` or `null`.
+		$loading_attrs['loading'] = 'lazy';
+	}
 
-		// Increase the counter since this is a main query content element.
-		$content_media_count = wp_increase_content_media_count();
+	/*
+	 * If flag was set based on contextual logic above, increase the content
+	 * media count if the image size is larger than the threshold.
+	 */
+	if ( $maybe_increase_count ) {
+		/** This filter is documented in wp-admin/includes/media.php */
+		$wp_min_priority_img_pixels = apply_filters( 'wp_min_priority_img_pixels', 50000 );
 
-		// If the count so far is below the threshold, `loading` attribute is omitted.
-		if ( $content_media_count <= wp_omit_loading_attr_threshold() ) {
-			// The first largest image will still get `fetchpriority='high'`.
-			return $postprocess( $loading_attrs, true );
+		if ( $wp_min_priority_img_pixels <= $attr['width'] * $attr['height'] ) {
+			wp_increase_content_media_count();
 		}
 	}
 
-	// Lazy-load by default for any unknown context.
-	$loading_attrs['loading'] = 'lazy';
-	return $postprocess( $loading_attrs, false );
+	// Potentially strip `loading="lazy"` if the feature is disabled.
+	if ( isset( $loading_attrs['loading'] ) && ! wp_lazy_loading_enabled( $tag_name, $context ) ) {
```

Comment:
It's a bit confusing to add the lazy loading attribute only to strip it at the end. Can we move the check for this being enabled earlier in the process and only add loading attributes when we've already determined that it's enabled? Additionally, I would image that if somehow someone programmatically added lazy loading to an image, this could remove that attribute. Which would be unexpected.

Reply:
That's a good point, I'll reverse this to only add the attribute if lazy-loading is enabled. This clause was like that before, so I probably just copied it over as is.

That said, unsetting the attribute even if it was manually provided would not lead to any functional bug, since in that case the `loading` attribute would already be within `$attr`, and then merging in the function return value into that with `array_merge()` would not drop the existing attribute. This function just returns it for consistency sake: Basically, the function will only ever return `loading="lazy"` or `fetchpriority="high"`, but if one of those two attributes is already provided in `$attr`, it's technically redundant and the function might as well return an empty array. Still, it returns those for consistency as those are what WordPress core would generally return.